### ISSUE TITLE
Future LRScheduler implementation with Tensor support

### ISSUE THREAD

**Issue by GdoongMathew** [pytorch/pytorch#128046]
### 🚀 The feature, motivation and pitch

Future `LRScheduler` implementation goals
1. we want to maintain the same semantics as when the lr was a Python number and not a Tensor:
    a. no inplacing -> we should not have any `fill_` in our LRScheduler math/compute. Instead we should assign with `group["lr"] = ...`.
    b. this goes hand-in-hand with cloning a copy during computation to assign back to the optimizer to avoid ever tangling the values.
2. we want to ensure that the LRScheduler preserves the "Tensorness" of the optimizer's hyperparams. In other words, if the Optimizer has a Tensor LR, the LRScheduler should always be giving it a Tensor LR to work with.

_Originally posted by @janeyx99 in https://github.com/pytorch/pytorch/issues/126854#issuecomment-2139963804_

**Comment by vadimkantorov** [comment]
This is a bit related to https://github.com/pytorch/pytorch/issues/68332 and https://github.com/pytorch/pytorch/issues/110636. In the sense that there is a more general problem that a pytorch function currently doesn't work nicely with Python scalars. In https://github.com/pytorch/pytorch/issues/110636 there was a proposal to support them as arguments to more pytorch functions (like torch.sqrt(x) currently does not accept Python scalars, but x ** 0.5 does), but make them always return tensors (I think this actually is what NumPy does, so this complicates partly the lanscape). Originally I proposed that many simple PyTorch functions should also support Python scalars but preserve tensorness and return a Python scalar if needed.

So in a way, I wonder does this preserve-tensorness-idea bring closer or further the purely functional scheduler formulas

**Comment by GdoongMathew** [comment]
Personally, I think either way would be fine, but supporting both tensor and pure scalar in the functional scheduler would be more challenging. Seems like fully converting to `torch.Tensor` before passing to functional scheduler would be one of the solution that solve this issue, and keep `LRScheduler`  from being too complex.

**Comment by GdoongMathew** [comment]
Hi @janeyx99 
I'd like to discuss a bit about the end goal here, with regard to the issue #130193. If I understand the comment https://github.com/pytorch/pytorch/issues/126854#issuecomment-2139963804  correctly, eventually we want torch to support tensor argument where its possible, but this would also cause issues, such as #130193, and #126854, where people might get confused of how tensor works in different classes / modules. 

Is this a legit design problem? or  the behavior is expected, since one is handled by an `nn.Module` and the other is by `optimizer & scheduler`, thus we should still keep the same principle as you mentioned, no `in-place` operation in `optimizer & scheduler`.

**Comment by janeyx99** [comment]
@GdoongMathew Your understanding for Optimizer + LRScheduler is correct, and the no-inplace expectation is still the way we want to go for LRScheduler integration with Optimizer. In the end, we don't expect LRScheduler + Optimizer to change mathematically if a hyperparameter is a Tensor vs a Python number.
